File: MyDailyBeat/MyDailyBeat/EVCGroupViewController.m
//
//  EVCGroupViewController.m
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 9/25/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//

#import "EVCGroupViewController.h"

@interface EVCGroupViewController ()

@end

@implementation EVCGroupViewController

- (void) EVCGroupSettingsViewControllerDelegateDidDeleteGroup:(EVCGroupSettingsViewController *) controller {
    [controller dismissViewControllerAnimated:YES completion:nil];
    EVCViewController *newcontroller = [[EVCViewController alloc] initWithNibName:@"EVCViewController_iPhone" bundle:nil];
    [newcontroller.view makeToast:@"Delete successful!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"check.png"]];
    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:newcontroller] animated:YES];
    
}

- (id) initWithGroup:(Group *) g andParent:(UIViewController *) parent {
    self = [self initWithNibName:@"EVCGroupViewController_iPhone" bundle:nil];
    if (self) {
        self.group = g;
        self.parentController = parent;
    }
    return self;
}

- (void) refreshGroupData {
    
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
            self.group.posts = [[RestAPI getInstance] getPostsForGroup:self.group];
            [self.view hideToastActivity];
        });
    });
}

- (void)writePost {
    EVCComposeViewControllerCompletionHandler completionHandler = ^(NSString* message, UIImage* image) {
        UIImage *attachedImage = image;
        NSString *postText = message;
        long long millis = [[NSDate date] timeIntervalSince1970];
        Post *written = [[Post alloc] init];
        written.postText = postText;
        written.dateTimeMillis = millis;
        written.userScreenName = [[RestAPI getInstance] getCurrentUser].screenName;
        dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
        dispatch_async(queue, ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view makeToastActivity];
            });
            NSData *imgData = UIImagePNGRepresentation(attachedImage);
            
            NSString *fileName = ASSET_FILENAME;
            
            BOOL success = [[RestAPI getInstance] writePost:written withPictureData:imgData andPictureName:fileName toGroup:self.group];
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view hideToastActivity];
                if (success)
                    [self.view makeToast:@"Upload successful!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"check.png"]];
                else {
                    [self.view makeToast:@"Upload failed!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"error.png"]];
                    return;
                }
                [self refreshGroupData];
                [self loadPosts];
            });
            
        });
        [self dismissModalViewControllerAnimated:YES];
    };
    EVCComposeViewController *composeNewPost = [[EVCComposeViewController alloc] initWithPostBlock:completionHandler];
    composeNewPost.title = @"New Post";
    self.modalPresentationStyle = UIModalPresentationFullScreen;
    
    [self presentModalViewController:[[UINavigationController alloc] initWithRootViewController:composeNewPost] animated:YES];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.scroll.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"texture.png"]];
    if (self.scroll.bounds.size.height >= max_post_height) {
        self.scroll.contentSize = self.scroll.bounds.size;
    } else {
       self.scroll.contentSize = CGSizeMake(self.scroll.bounds.size.width, max_post_height+10);
    }
    
    self.scroll.alwaysBounceVertical = YES;
    self.imageView.layer.shadowColor = [UIColor purpleColor].CGColor;
    self.imageView.layer.shadowOffset = CGSizeMake(0, 1);
    self.imageView.layer.shadowOpacity = 1;
    self.imageView.layer.shadowRadius = 1.0;
    self.imageView.clipsToBounds = YES;
    
     _composeButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCompose target:self action:@selector(writePost)];
    
    UIBarButtonItem *space = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
    
    _inviteButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"add_user-25.png"] style:UIBarButtonItemStylePlain target:self action:@selector(invite)];
    
    _settingsButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"settings-25.png"] style:UIBarButtonItemStylePlain target:self action:@selector(groupSettings)];
    
    NSArray *items = [[NSArray alloc] initWithObjects:_composeButton, space, _inviteButton, _settingsButton, nil];
    [self.groupBar setItems:items animated:YES];
    
    
    self.title = self.group.groupName;
    
    UIImage* image3 = [EVCCommonMethods imageWithImage:[UIImage imageNamed:@"hamburger-icon-white"] scaledToSize:CGSizeMake(30, 30)];
    CGRect frameimg = CGRectMake(0, 0, image3.size.width, image3.size.height);
    UIButton *someButton = [[UIButton alloc] initWithFrame:frameimg];
    [someButton setBackgroundImage:image3 forState:UIControlStateNormal];
    [someButton addTarget:self action:@selector(showMenu)
         forControlEvents:UIControlEventTouchUpInside];
    [someButton setShowsTouchWhenHighlighted:YES];
    
    UIBarButtonItem *menuButton =[[UIBarButtonItem alloc] initWithCustomView:someButton];
    
    self.navigationItem.rightBarButtonItem = menuButton;
    
    UIImage* image4 = [EVCCommonMethods imageWithImage:[UIImage imageNamed:@"profile-icon-white"] scaledToSize:CGSizeMake(30, 30)];
    CGRect frameimg2 = CGRectMake(0, 0, image4.size.width, image4.size.height);
    UIButton *someButton2 = [[UIButton alloc] initWithFrame:frameimg2];
    [someButton2 setBackgroundImage:image4 forState:UIControlStateNormal];
    [someButton2 addTarget:self action:@selector(showProfile)
          forControlEvents:UIControlEventTouchUpInside];
    [someButton2 setShowsTouchWhenHighlighted:YES];
    
    UIBarButtonItem *profileButton =[[UIBarButtonItem alloc] initWithCustomView:someButton2];
    self.navigationItem.leftBarButtonItem = profileButton;
    
    if ([[[RestAPI getInstance] getCurrentUser].screenName isEqualToString:self.group.adminName]) {
        [_settingsButton setEnabled:YES];
    } else {
        [_settingsButton setEnabled:NO];
    }
    
    [self refreshGroupData];
    [self loadPicture];
    [self loadPosts];
    
    
}

- (void) invite {
    EVCUserSearchViewViewController *searchController = [[EVCUserSearchViewViewController alloc] init];
    searchController.groupToInviteTo = self.group;
    [self presentViewController:[[UINavigationController alloc] initWithRootViewController:searchController] animated:YES completion:nil];

}

- (void) loadPicture {
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        NSURL *imageURL = [[RestAPI getInstance] retrieveGroupPictureForGroup:self.group];
        if (imageURL == nil) return;
        NSData *imageData = [[RestAPI getInstance] fetchImageAtRemoteURL:imageURL];
        dispatch_async(dispatch_get_main_queue(), ^{
            // Update the UI
            [self.view hideToastActivity];
            self.imageView.image = [UIImage imageWithData:imageData];
            
        });
    });
    
}

- (void) deletePost:(Post *) p {
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        
        BOOL success = [[RestAPI getInstance] deletePost:p];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideToastActivity];
            if (success)
                [self.view makeToast:@"Delete successful!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"check.png"]];
            else {
                [self.view makeToast:@"Delete failed!" duration:3.5 position:@"bottom" image:[UIImage imageNamed:@"error.png"]];
                return;
            }
            [self refreshGroupData];
            [self loadPosts];
        });
        
    });
}

- (void)groupSettings {
    EVCGroupSettingsViewController *settingsController = [[EVCGroupSettingsViewController alloc] initWithGroup:self.group andCompletionBlock:^{
        [self refreshGroupData];
        [self loadPicture];
        [self loadPosts];
    }];
    settingsController.delegate = self;
    [settingsController setTitle:@"Group Settings"];
    [self presentViewController:[[UINavigationController alloc] initWithRootViewController:settingsController] animated:YES completion:nil];
}

- (void) loadPosts {
    NSArray *viewsToRemove = [self.scroll subviews];
    for (UIView *v in viewsToRemove) {
        [v removeFromSuperview];
    }
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        [self refreshGroupData];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideToastActivity];
            [self.group.posts sortUsingComparator:^NSComparisonResult(id obj1, id obj2) {
                
                int id1 = [(Post *)obj1 post_id];
                int id2 = [(Post *)obj2 post_id];
                return id1 > id2;
            }];
            max_post_height = 10;
            int count = (int)[self.group.posts count];
            for (int i = count - 1 ; i >= 0 ; --i) {
                [self.scroll addSubview:[self createViewForPost:[self.group.posts objectAtIndex:i]]];
            }
            self.scroll.contentSize = CGSizeMake(self.scroll.bounds.size.width, max_post_height+10);
        });
    });
    
}

- (UIView *) createViewForPost: (Post *) inputPost {
    UIView *postView;
    [postView setBackgroundColor:[UIColor whiteColor]];
    if (inputPost.blobKey != nil) {
        EVCPostView *post = [[EVCPostView alloc] initWithFrame:CGRectMake(0, 0, 300, 400) andPost:inputPost withPostType:EVCPostTypeHasPicture andParent:self];
        postView = [[UIView alloc] initWithFrame:CGRectMake(10, max_post_height, 300, 400)];
        [postView addSubview:post];
        
        max_post_height += 430;
        
        
    } else {
        EVCPostView *post = [[EVCPostView alloc] initWithFrame:CGRectMake(0, 0, 300, 178) andPost:inputPost withPostType:EVCPostTypeDoesNotHavePicture andParent:self];
        postView = [[UIView alloc] initWithFrame:CGRectMake(10, max_post_height, 300, 178)];
        [postView addSubview:post];
        
        max_post_height += 208;
        
    }
    
    // drop shadow
    [postView.layer setShadowColor:[UIColor blackColor].CGColor];
    [postView.layer setShadowOpacity:0.8];
    [postView.layer setShadowRadius:3.0];
    [postView.layer setShadowOffset:CGSizeMake(2.0, 2.0)];
    
    return postView;
    
}

- (void) showMenu {
    [self.sideMenuViewController presentRightMenuViewController];
}

- (void) showProfile {
    [self.sideMenuViewController presentLeftMenuViewController];
}

@end

File: MDB_Swift/MyDailyBeat/MyDailyBeat/src/EVCGroupViewController.swift

//
//  EVCGroupViewController.swift
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 9/25/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//
import UIKit
import QuartzCore
import Toast_Swift
import API
class EVCGroupViewController: UIViewController {
    var max_post_height: Int = 0

    @IBOutlet var scroll: UIScrollView!
    @IBOutlet var imageView: UIImageView!
    var composeButton: UIBarButtonItem!
    var settingsButton: UIBarButtonItem!
    var inviteButton: UIBarButtonItem!
    var group: Group!
    @IBOutlet var groupBar: UIToolbar!
    var parentController: UIViewController!

    convenience init(group g: Group, andParent parent: UIViewController) {
        self.init(nibName: "EVCGroupViewController_iPhone", bundle: nil)
        
        self.group = g
        self.parentController = parent
    
    }
    
    override var nibName: String? {
        return "EVCGroupViewController_iPhone"
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.scroll.backgroundColor = UIColor.groupTableViewBackground
        if self.scroll.bounds.size.height >= CGFloat(max_post_height) {
            self.scroll.contentSize = self.scroll.bounds.size
        }
        else {
            self.scroll.contentSize = CGSize(width: CGFloat(self.scroll.bounds.size.width), height: CGFloat(max_post_height + 10))
        }
        self.scroll.alwaysBounceVertical = true
        self.imageView?.layer.shadowColor = UIColor.purple.cgColor
        self.imageView?.layer.shadowOffset = CGSize(width: CGFloat(0), height: CGFloat(1))
        self.imageView?.layer.shadowOpacity = 1
        self.imageView?.layer.shadowRadius = 1.0
        self.imageView?.clipsToBounds = true
        self.composeButton = UIBarButtonItem(barButtonSystemItem: .compose, target: self, action: #selector(self.writePost))
        let space = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
        self.inviteButton = UIBarButtonItem(image: UIImage(named: "add_user-25.png"), style: .plain, target: self, action: #selector(self.invite))
        self.settingsButton = UIBarButtonItem(image: UIImage(named: "settings-25.png"), style: .plain, target: self, action: #selector(self.groupSettings))
        let items = [self.composeButton!, space, self.inviteButton!, self.settingsButton!]
        self.groupBar.setItems(items, animated: true)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.setNavTitle(to: self.group.groupName)
        
        if (RestAPI.getInstance().getCurrentUser().id == self.group.adminID) {
            self.settingsButton.isEnabled = true
            self.inviteButton.isEnabled = true
        } else {
            self.settingsButton.isEnabled = false
            self.inviteButton.isEnabled = false
        }
        self.refreshGroupData()
        self.loadPicture()
        self.loadPosts()
    }

    func writePost() {
        let completionHandler: (_ message: String, _ image: UIImage?) -> Void = { (message, image) in
            let attachedImage: UIImage? = image
            let postText: String = message
            let millis: Int64 = Int64(Date().timeIntervalSince1970)
            let written = Post()
            written.postText = postText
            written.dateTimeMillis = millis
            written.userid = RestAPI.getInstance().getCurrentUser().id
            
            DispatchQueue.global().async(execute: {() -> Void in
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.makeToastActivity(ToastPosition.center)
                })
                let imgData: Data?
                if let attachment = attachedImage {
                    imgData = UIImagePNGRepresentation(attachment)
                } else {
                    imgData = nil
                }
                let fileName: String = ASSET_FILENAME
                let success: Bool = RestAPI.getInstance().write(written, withPictureData: imgData, andPictureName: fileName, to: self.group)
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.hideToastActivity()
                    if success {
                        self.view.makeToast("Upload successful!", duration: 3.5, position: .bottom)
                    }
                    else {
                        self.view.makeToast("Upload failed!", duration: 3.5, position: .bottom)
                        return
                    }
                    self.refreshGroupData()
                    self.loadPosts()
                })
            })
            self.dismiss(animated: true, completion: nil)
        }
        let composeNewPost = EVCComposeViewController()
        composeNewPost.completionHandler = completionHandler
        composeNewPost.title = "New Post"
        self.modalPresentationStyle = .fullScreen
        self.present(UINavigationController(rootViewController: composeNewPost), animated: true, completion: nil)
    }

    func groupSettings() {
        self.performSegue(withIdentifier: "EditGroupSegue", sender: nil)
    }

    func deletePost(_ p: Post) {
        
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(ToastPosition.center)
            })
            let success: Bool = RestAPI.getInstance().delete(post: p)
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
                if success {
                    self.view.makeToast("Delete successful!", duration: 3.5, position: .bottom)
                }
                else {
                    self.view.makeToast("Delete failed!", duration: 3.5, position: .bottom)
                    return
                }
                self.refreshGroupData()
                self.loadPosts()
            })
        })
    }

    func refreshGroupData() {
        
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(ToastPosition.center)
                self.group.posts = RestAPI.getInstance().getPostsFor(self.group)
                self.view.hideToastActivity()
            })
        })
    }

    func invite() {
        let searchController = EVCUserSearchViewViewController()
        searchController.groupToInviteTo = self.group
        self.present(UINavigationController(rootViewController: searchController), animated: true, completion: { _ in })
    }

    func loadPicture() {
        
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(ToastPosition.center)
            })
            let imageURL: URL? = RestAPI.getInstance().retrieveGroupPicture(for: self.group)
            if imageURL == nil {
                return
            }
            let imageData: Data? = RestAPI.getInstance().fetchImage(atRemoteURL: imageURL!)
            DispatchQueue.main.async(execute: {() -> Void in
                // Update the UI
                self.view.hideToastActivity()
                self.imageView.image = UIImage(data: imageData!)
            })
        })
    }

    func loadPosts() {
        let viewsToRemove: [UIView] = self.scroll.subviews
        for v: UIView in viewsToRemove {
            v.removeFromSuperview()
        }
        
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(ToastPosition.center)
            })
            self.refreshGroupData()
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
                self.group.posts.sort(by: {(_ obj1: Post, _ obj2: Post) -> Bool in
                    let id1: Int = obj1.post_id
                    let id2: Int = obj2.post_id
                    return id1 > id2
                })
                self.max_post_height = 10
                let count = Int(self.group.posts.count)
                var i = count - 1
                while i >= 0 {
                    self.scroll.addSubview(self.createView(for: self.group.posts[i])!)
                    i -= 1
                }
                self.scroll.contentSize = CGSize(width: CGFloat(self.scroll.bounds.size.width), height: CGFloat(self.max_post_height + 10))
            })
        })
    }

    func createView(for inputPost: Post) -> UIView? {
        let postView: UIView?
        if inputPost.imageUrl != "" {
            let post = EVCPostView(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(300), height: CGFloat(400)), andPost: inputPost, with: EVCPostType.hasPicture, andParent: self)
            postView = UIView(frame: CGRect(x: CGFloat(10), y: CGFloat(max_post_height), width: CGFloat(300), height: CGFloat(400)))
            postView?.addSubview(post)
            max_post_height += 430
        }
        else {
            let post = EVCPostView(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(300), height: CGFloat(178)), andPost: inputPost, with: EVCPostType.doesNotHavePicture, andParent: self)
            postView = UIView(frame: CGRect(x: CGFloat(10), y: CGFloat(max_post_height), width: CGFloat(300), height: CGFloat(178)))
            postView?.addSubview(post)
            max_post_height += 208
        }
        postView?.backgroundColor = UIColor.white
        // drop shadow
        postView?.layer.shadowColor = UIColor.black.cgColor
        postView?.layer.shadowOpacity = 0.8
        postView?.layer.shadowRadius = 3.0
        postView?.layer.shadowOffset = CGSize(width: CGFloat(2.0), height: CGFloat(2.0))
        return postView!
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let dest = segue.destination as? EVCGroupSettingsViewController {
            dest.group = self.group
            self.navigationController?.setNavigationBarHidden(false, animated: true)
        }
    }

    
}
